run/observer.ts

```diff
@@ -1,28 +1,53 @@
 import type { Observer, Resolve } from "../types.ts";
 
-import { shift } from "../deps.ts";
+import { Computation, shift } from "../deps.ts";
 
 export function createObservable<T>() {
   let observers = new Map<Observer<T>, Resolve<T>>();
 
-  return {
+  let observable = {
     notify(value: T) {
-      for (let listener of observers.values()) {
-        listener(value);
+      let handlers = [...observers.values()];
+      for (let handler of handlers) {
+        handler(value);
       }
     },
+    *first(): Computation<T> {
+      return yield* shift<T>(function* (k) {
+        let observer = observable.observe();
+        let value = yield* observer;
+        observer.drop();
+        k(value);
+      });
+    },
     observe(): Observer<T> {
+      let events: T[] = [];
+      let consumers: Resolve<T>[] = [];
       let observer = {
         *[Symbol.iterator]() {
-          return yield* shift<T>(function* (k) {
-            observers.set(observer, k);
-          });
+          let event = events.pop();
+          if (event) {
+            return event;
+          } else {
+            return yield* shift<T>(function* (k) {
+              consumers.unshift(k);
+            });
+          }
         },
         drop() {
           observers.delete(observer);
         },
       };
+      observers.set(observer, (event: T) => {
+        events.unshift(event);
+        while (events.length > 0 && consumers.length > 0) {
+          let consume = consumers.pop() as Resolve<T>;
+          let event = events.pop() as T;
+          consume(event);
+        }
+      });
       return observer;
     },
   };
+  return observable;
 }
```

Comment:
Is this so the observer is added immediately instead of the next tick and can start receiving events?

Reply:
I think so yes. It's so that you can add the observer, and if any events are dispatched at any point before you actually want to get the next one, you don't drop any of them.